### PR TITLE
fix coverity defects with CID 147626-28

### PR DESCRIPTION
**Pull request by heary-cao**
issues:
fix coverity defects
coverity scan CID:147626, Type:Dereference before null check
coverity scan CID:147628, Type:Dereference before null check

Signed-off-by: cao.xuewen cao.xuewen@zte.com.cn

### CODE REVIEW

File: module/zfs/zfs_vnops.c [hunk 1]
```diff
@@ -4090,8 +4090,9 @@ zfs_putpage(struct inode *ip, struct page *pp, struct writeback_control *wbc)
 
 	err = sa_bulk_update(zp->z_sa_hdl, bulk, cnt, tx);
 
-	zfs_log_write(zsb->z_log, tx, TX_WRITE, zp, pgoff, pglen, 0,
-	    zfs_putpage_commit_cb, pp);
+	if (zsb->z_log != NULL)
+		zfs_log_write(zsb->z_log, tx, TX_WRITE, zp, pgoff, pglen, 0,
+		    zfs_putpage_commit_cb, pp);
```
Comment (behlendorf) [inline module/zfs/zfs_vnops.c]: This should never be a possible case in the existing code.   Pages which are part of a snapshot should never be allowed to be dirtied, see commit 1e8db7710220332808920a582e5794d6fc37b109.  This check was added just in case somehow a page got dirtied.  Now that pretty clearly can never happen or we would have NULL dereferenced here long ago and there are no issues like this open in the tracker.  Given that I'd rather remove the NULL check around `zil_commit()` a few lines down.

This also has the advantage of bringing this code slightly closer inline with upstream at least in terms of intent and assumptions.

@tuxoko what are your thoughts on this.

File: module/zfs/zfs_vfsops.c [hunk 2]
```diff
@@ -1701,8 +1701,7 @@ zfs_resume_fs(zfs_sb_t *zsb, const char *osname)
 		 * Since we couldn't setup the sa framework, try to force
 		 * unmount this file system.
 		 */
-		if (zsb->z_os)
-			(void) zfs_umount(zsb->z_sb);
+		(void) zfs_umount(zsb->z_sb);
```
Comment (behlendorf) [inline module/zfs/zfs_vfsops.c]: That's a good catch.

File: lib/libuutil/uu_dprintf.c [hunk 3]
```diff
@@ -68,7 +68,8 @@ uu_dprintf_create(const char *name, uu_dprintf_severity_t severity,
 {
 	uu_dprintf_t *D;
 
-	if (uu_check_name(name, UU_NAME_DOMAIN) == -1) {
+	if (name != NULL ||
+	    uu_check_name(name, UU_NAME_DOMAIN) == -1) {
```
Comment (tuxoko) [inline lib/libuutil/uu_dprintf.c]: What are you trying to do...

File: module/zfs/zfs_fm.c [hunk 4]
```diff
@@ -131,7 +131,8 @@ zfs_ereport_start(nvlist_t **ereport_out, nvlist_t **detector_out,
 	 * If we are doing a spa_tryimport() or in recovery mode,
 	 * ignore errors.
 	 */
-	if (spa_load_state(spa) == SPA_LOAD_TRYIMPORT ||
+	if (spa != NULL ||
+	    spa_load_state(spa) == SPA_LOAD_TRYIMPORT ||
```
Comment (tuxoko) [inline module/zfs/zfs_fm.c]: What are you trying to do....

File: module/zfs/zfs_vfsops.c (hunk 2 above)
Comment (tuxoko) [inline module/zfs/zfs_vfsops.c]: Can you explain this change?

File: module/zfs/zfs_vnops.c (hunk 1 above)
Comment (tuxoko) [inline module/zfs/zfs_vnops.c]: @behlendorf 
Yes, I agree with you. I don't see why we need this.

**Comment by tuxoko** [comment]
@behlendorf 
TBH, I really don't like a mechanical fix like this. Just as this patch shows, when you just try to silence a static analysis, but not really understand what the code you touch is doing. It often make things worse.

File: module/zfs/zfs_fm.c (hunk 4 above)
Comment (tuxoko) [inline module/zfs/zfs_fm.c]: @behlendorf 
If you look carefully, it doesn't fix null dereference...

Comment (behlendorf) [inline module/zfs/zfs_fm.c]: Right the fix itself is wrong here.  It should be `spa == NULL` in the check so it returns immediately return when no spa is passed.  Or better yet just drop think hunk completely.  This is an internal function all callers know to always pass a spa.  Period.

**Comment by behlendorf** [comment]
@tuxoko it's a useful tool as long as it's not used purely mechanically.  Each change needs to be understood and you're right it shouldn't be used just to silence a warning but to make the code clearer.

File: module/zfs/zfs_vfsops.c (hunk 2 above)
Comment (behlendorf) [inline module/zfs/zfs_vfsops.c]: The original fix here doesn't look quite right.  Take a look at the stack from the original bug report.  If zsb->z_os were NULL we should never have been able to make it to where the node finally crashed.

https://github.com/zfsonlinux/zfs/issues/1518

```
zfs_resume_fs
  zfs_umount
    zfs_sb_teardown
      dmu_objset_pool <- PANIC
```

In `zfs_sb_teardown` several checks for `zsb->z_os == NULL` are made before it's used which should have prevented this.  What exactly in this first check preventing that the other ones aren't?

Comment (tuxoko) [inline module/zfs/zfs_vfsops.c]: @behlendorf 
The commit message says concurrent umount, so it's likely caused by race. Though I don't think the patch actually fix the race. Also, using zfs_umount here looks strange, I'm not sure how things would work in terms of VFS after this call.

Comment (behlendorf) [inline module/zfs/zfs_vfsops.c]: The fact that zsb->z_os can change outside a lock is definitely racy and problematic.  Rather than needlessly change this let's leave it as is for now and I'll make a note in the Coverity analysis.  @heary-cao go ahead and drop this hunk when you next refresh this patch.

File: module/zfs/zfs_fm.c (hunk 4 above)
Comment (behlendorf) [inline module/zfs/zfs_fm.c]: This chunk isn't correct.  Yes it prevents a NULL dereference in this block but the same thing can happen 5 lines down.  Please drop think hunk from the patch and we'll address it if need be in a different PR.

File: lib/libuutil/uu_dprintf.c [hunk 5]
```diff
@@ -76,14 +76,10 @@ uu_dprintf_create(const char *name, uu_dprintf_severity_t severity,
...
-		if (D->uud_name == NULL) {
-			uu_free(D);
-			return (NULL);
-		}
-	} else {
-		D->uud_name = NULL;
+	D->uud_name = strdup(name);
+	if (D->uud_name == NULL) {
+		uu_free(D);
+		return (NULL);
```
Comment (behlendorf) [inline lib/libuutil/uu_dprintf.c]: The intention of this function is clearly to allow name to be NULL.  Rather than removing this code I'd recommend only running the `uu_check_name()` when `name != NULL`.  I'd suggest moving to it's own PR too so we can merge the other parts of this PR.

Comment (heary-cao) [inline lib/libuutil/uu_dprintf.c]: @behlendorf
thanks for your suggest
commit other PR:https://github.com/zfsonlinux/zfs/pull/5376
please review it.
thanks.

**Comment by behlendorf** [comment]
@tuxoko any remaining concerns with the trimmed down patch proposed.
